### PR TITLE
Hw8 t02

### PR DESCRIPTION
**Pull request by mirmozavr**
TableData for database table

### CODE REVIEW

File: hw8/hw8_t02.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+Assume all data has unique values in 'name' column.
+So, if presidents = TableData(database_name='example.sqlite', table_name='presidents')
+
+then
+ -  `len(presidents)` will give current amount of rows in presidents table in database
+ -  `presidents['Yeltsin']` should return single data row for president with name Yeltsin
+ -  `'Yeltsin' in presidents` should return if president with same name exists in table
+ -  object implements iteration protocol. i.e. you could use it in for loops::
+       for president in presidents:
+           print(president['name'])
```
Comment (mirmozavr) [inline hw8/hw8_t02.py:15]: What's the right way to test this requirement?

File: hw8/hw8_t02.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        self.cursor = self.connect_and_return_cursor()
+        self.cursor.execute(f"SELECT count(*) FROM {self.table_name[0]}")  # noqa: S608
+        return self.cursor.fetchone()[0]
+
+    def __iter__(self):
+        self.cursor = self.connect_and_return_cursor()
+        for row in self.cursor.execute(
+            f"SELECT * FROM {self.table_name[0]}"  # noqa: S608
+        ):  # noqa: S608
+            yield {"name": row[0], "age": row[1], "country": row[2]}
```
Comment (mirmozavr) [inline hw8/hw8_t02.py]: Invented to yield a dict to fit 
`       for president in presidents:
           print(president['name'])`
requirement...

Comment (mirmozavr) [reply hw8/hw8_t02.py]: I just got familiar with `sqlite3.Row` objects! Need a lil more time

File: hw8/hw8_t02.py (hunk 1 above)
Comment (Cjkjvfnby) [reply hw8/hw8_t02.py:15]: This is the requirement: 
```
object implements iteration protocol
```

This is the example:
```python
# i.e. you could use it in for loops
for president in presidents:
           print(president['name'])
```

File: hw8/hw8_t02.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ Use supplied example.sqlite file as database fixture file.
+"""
+
+import sqlite3
+
+
+class TableData:
+    def __init__(self, db_path: str, table_name: str):
+        self.db_path = db_path
+        self.table_name = (table_name,)
```
Comment (Cjkjvfnby) [inline hw8/hw8_t02.py]: Why?

File: hw8/hw8_t02.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        for row in self.cursor.execute(
+            f"SELECT * FROM {self.table_name[0]}"  # noqa: S608
+        ):  # noqa: S608
+            yield row
+
+    def __getitem__(self, item: str):
+        item = (item,)
+        self.cursor = self.connect_and_return_cursor()
+        self.cursor.execute(
+            f"SELECT * FROM {self.table_name[0]} WHERE name = '{item[0]}'"  # noqa: S608
```
Comment (Cjkjvfnby) [inline hw8/hw8_t02.py]: Why you have suppressed S608?

File: hw8/hw8_t02.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        self.cursor = self.connect_and_return_cursor()
+        self.cursor.execute(f"SELECT count(*) FROM {self.table_name[0]}")  # noqa: S608
+        return self.cursor.fetchone()[0]
+
+    def __iter__(self):
+        self.cursor = self.connect_and_return_cursor()
+        for row in self.cursor.execute(
+            f"SELECT * FROM {self.table_name[0]}"  # noqa: S608
+        ):  # noqa: S608
+            yield row
```
Comment (Cjkjvfnby) [inline hw8/hw8_t02.py]: You could use yield from
```suggestion
       yield from self.cursor.execute(
            f"SELECT * FROM {self.table_name[0]}"  # noqa: S608
        )  # noqa: S608
```

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
If you open a connection what should you also do?

**Comment by mirmozavr** [comment]
I'm trying to follow guidelines. To prevent SQL-injection. And S608 is appearing anyway :(

# Never do this -- insecure!
symbol = 'RHAT'
cur.execute("SELECT * FROM stocks WHERE symbol = '%s'" % symbol)

# Do this instead
t = ('RHAT',)
cur.execute('SELECT * FROM stocks WHERE symbol=?', t)
print(cur.fetchone())

**Comment by mirmozavr** [comment]
> 
> 
> If you open a connection what should you also do?

It should be closed. I came with this solution:

```
def some_func(self):
    try:
         connection...
         return results...
    finally:
        connection.close()
```
Doesn't look nice, to be honest

Or
`with sqlite3.connection..... as conn:`

**Comment by mirmozavr** [comment]
> with sqlite3.connection..... as conn:

Well, `with` doesn't close the connection afterward.

**Comment by mirmozavr** [comment]
Discovered an idea: make my `TableData` class a context manager.
Do all connections and cursors in `__enter__` and all closures in `__exit__`

**Comment by Cjkjvfnby** [comment]
> Well, with doesn't close the connection afterward.

Yes, it is unexpected. https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager

See https://docs.python.org/3/library/contextlib.html#contextlib.closing

**Comment by Cjkjvfnby** [comment]
> Discovered an idea: make my TableData class a context manager. Do all connections and cursors in __enter__ and all closures in __exit__

This was my next question :)

File: hw8/hw8_t02.py (hunk 3 above)
Comment (Cjkjvfnby) [reply hw8/hw8_t02.py]: Why tuple?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
> I'm trying to follow guidelines. To prevent SQL-injection. And S608 is appearing anyway :(

Please looks closer to `Do this instead` section. Is there any difference with your implementation?

PS. You can leave table substitution, it is hard to deal with it.

**Comment by mirmozavr** [comment]
> Why tuple?
Substitution needed an iterable, I think.

File: hw8/tests/test_hw8_t02.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            names.append(president["name"])
+            ages.append(president["age"])
+            countries.append(president["country"])
+        assert names == ["Yeltsin", "Trump", "Big Man Tyrone"]
+        assert ages == [999, 1337, 101]
+        assert countries == ["Russia", "US", "Kekistan"]
+
+
+def test_table_data_length():
+    presidents = TableData(db_path, "presidents")
```
Comment (Cjkjvfnby) [inline hw8/tests/test_hw8_t02.py:37]: Do you close the file?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: hw8/tests/test_hw8_t02.py (hunk 6 above)
Comment (mirmozavr) [reply hw8/tests/test_hw8_t02.py:37]: Nope. Thanks! Added a method for it.

File: hw8/hw8_t02.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        return tuple(self.cursor.fetchone())
+
+    def __contains__(self, item: str):
+        self.cursor.execute(
+            f"SELECT count(*) FROM {self.table_name} WHERE name = ?",  # noqa: S608
+            (item,),
+        )
+        return self.cursor.fetchone()[0] > 0
+
+    def close(self) -> None:
```
Comment (Cjkjvfnby) [inline hw8/hw8_t02.py:66]: Actually, I prefer a solution when you could only call this class via context manager.  So the user will never forget to close it. 

There are a lot of Python classes that allow both approaches, but most of them were designed before context manager and preserve backward compatibility.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**